ratpack-pac4j/src/main/java/ratpack/pac4j/RatpackPac4j.java

```diff
@@ -496,4 +508,21 @@ private static void initiateAuthentication(Context ctx, Class<? extends Client<?
       webContext.sendResponse();
     });
   }
+
+  private static Promise<Optional<UserProfile>> performDirectAuthentication(Context ctx, Class<? extends Client<?, ?>> clientType) {
+    return internalWebContext(ctx).flatMap(webContext ->
+      Blocking
+        .get(() -> {
+          Clients clients = ctx.get(Clients.class);
+          Client<?, ?> client = clients.findClient(clientType);
+          return userProfileFromCredentials(client, webContext);
+        })
+    );
+  }
+
+  private static <C extends Credentials, U extends UserProfile> Optional<UserProfile> userProfileFromCredentials(Client<C, U> client, RatpackWebContext webContext) throws RequiresHttpAction {
```

Comment:
Trying to inline this method just got me tangled up in generic type resolution hell. If I add these same generic types to the `performDirectAuthentication` method I need to propagate them right up to the call site.